### ISSUE TITLE
S3 Sink - failed request not retried due to IMDS timeout

### ISSUE THREAD

**Issue by unkempthenry** [vectordotdev/vector#14671]
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Problem

We're currently seeing intermittent failures of the S3 sink that are not retried.  The cause appears to be that in our env it occasionally takes >1 second for IMDS to respond with credentials ([this 1 second timeout is set in the AWS Rust SDK](https://github.com/awslabs/aws-sdk-rust/blob/9bf3d6f46df1ea5c1f3116f5942c08de95366ac8/sdk/aws-config/src/imds/client.rs#L52)).  A complicating factor is that we are also running vector on a K8s cluster and using KIAM to intercept the requests which is likely exacerbating the issue.

Since these IMDS responses can intermittently take longer than 1 second, it would be nice to be able to either:
1) retry these requests
2) configure a longer timeout

Here's an example error:
```
{"error":"failed to construct request: Failed to load credentials from the credentials provider: An error occurred while loading credentials: An IO error occurred communicating with IMDS: timeout: HTTP read timeout occurred after 1s","message":"Non-retriable error; dropping the request.","metadata":{"kind":"event","level":"ERROR","module_path":"vector::sinks::util::retries","target":"vector::sinks::util::retries"},"pid":8,"timestamp":"2022-09-30T18:44:52.996691840Z","vector":{"component_id":"s3_logs_sink","component_kind":"sink","component_name":"s3_logs_sink","component_type":"aws_s3"}}
```

### Configuration

```text
[sinks.s3_logs_sink]
      type = "aws_s3"
      inputs = [ "to_s3" ]
      bucket = "..."
      key_prefix = "{{ s3_root_key }}/year=%Y/month=%m/day=%d/hour=%H/minute=%M/"
      filename_extension = "ndjson"
      compression = "gzip"
      region = "us-east-1"
      batch.max_bytes = 10000000 # 10 MB
      batch.timeout_secs = 30
      encoding.except_fields = [...]
      encoding.codec = "native_json"
```

### Version

vector 0.24.1 (x86_64-unknown-linux-gnu 8935681 2022-09-12)

### Additional Context

Vector is running on Kubernetes and IMDS requests are being intercepted by KIAM.  We're currently trying to diagnose why KIAM is taking >1 second to return the IMDS response, however since this problem can happen intermittently, it feels like something that should be retried.

### References

We've created a [feature request](https://github.com/awslabs/aws-sdk-rust/issues/625) on the SDK project to support configuring the IMDS timeout via env vars.

**Comment by unkempthenry** [comment]
Looks like the AWS Rust SDK maintainers are on the fence about allowing the IMDS timeout and retry behavior to be configured via Environment Variables: https://github.com/awslabs/smithy-rs/pull/1821#issuecomment-1273607602 .

@jszwedko , do you have an opinion on what a solution in Vector should look like? Maybe it's sufficient to just allow this request to be retried?  Alternatively, these options could probably be exposed as options for the S3 sink configuration, but I think it would make the sink a bit more complicated (would need to create the client with non-default settings).

**Comment by jszwedko** [comment]
> Looks like the AWS Rust SDK maintainers are on the fence about allowing the IMDS timeout and retry behavior to be configured via Environment Variables: [awslabs/smithy-rs#1821 (comment)](https://github.com/awslabs/smithy-rs/pull/1821#issuecomment-1273607602) .
> 
> @jszwedko , do you have an opinion on what a solution in Vector should look like? Maybe it's sufficient to just allow this request to be retried? Alternatively, these options could probably be exposed as options for the S3 sink configuration, but I think it would make the sink a bit more complicated (would need to create the client with non-default settings).

Interesting. We could just reuse the existing `request` options available on HTTP-based sinks, like `aws_s3`. It would just mean applying them to the IMDS requests too (and probably only the timeout and retry options, not the concurrency options). We'd be happy to see a PR for that.

**Comment by kevinpark1217** [comment]
I created a PR #15518 which allows IMDS requests retry & timeouts to be configurable. I think this will allow @unkempthenry and I to navigate around the flaky IMDS proxy.

I was able to convince the AWS Rust SDK maintainers to at least make the IMDS client in the default credentials chain overridable, which was released with version `0.51.0`. More information on https://github.com/awslabs/smithy-rs/pull/1867
